### ISSUE TITLE
Cyclic reference and ShouldBeEquivalentTo - is this a valid solution?

### ISSUE THREAD

**Issue by fnajera-rac-de** [dennisdoomen/FluentAssertions#157]
Hi,

I want to deep-test two objects for equivalency, but they contain cycles so the library halts with "but it contains a cyclic reference.". I know it is possible to ignore those properties, but I want them tested.

The solution I have found is:

* EquivalencyValidator: remove the line

        if (!current.Get<ObjectTracker>("objects").IsCyclicReference(new ObjectReference(context.Subject, context.PropertyPath)))

* Create a new IAssertionRule class, and use it (surely this can use some refactoring / improvements):

        private class MyAssertionRule : IAssertionRule
        {

            private List<Tuple<object, object>> _list = new List<Tuple<object, object>>();

            public bool AssertEquality(IEquivalencyValidationContext context)
            {
                var exists = _list.Any(x => object.ReferenceEquals(x.Item1, context.Subject) && object.ReferenceEquals(x.Item2, context.Expectation));
                if (exists)
                {
                    return true;
                }
                else
                {
                    _list.Add(new Tuple<object, object>(context.Subject, context.Expectation));
                    return false;
                }
            }

        }

The reasoning is:

* The assertion rule is called for each pair of matching properties.

* By removing the "if" line, I don't fail on cycles but I don't ignore them either. Note that by doing so without adding the new rule you will certainly hit the maximum depth limit.

* In my rule, if it's the first time I see a given pair of values, then I cannot conclude if they match, so I return false, but I note down that they appeared before together.

    * If those values are indeed not equivalent, then the process will eventually stop with an error, so it finally won't matter that I noted them down.

    * If those values are found equivalent by another rule or by finishing the deep inspection, then they were indeed a match, and I have that match on my list, which connects with the following point.

* If I have seen this pair of values already, then I can return true, because either they were compared before and they were found to be equivalent - or - this specific pair of objects is being compared right now and this property specifies "the same cycle" in both objects.

* An important point is that any given pair of objects will be thoroughly tested once (because the first time I see them I return false).

Am I missing something here? If not, could this be a universal way of dealing with cycles - maybe applied by default?

Thanks and regards,

Fernando

**Comment by dennisdoomen** [comment]
@vossad01 This is still ongoing work, isn't it?

**Comment by vossad01** [comment]
@dennisdoomen I did not have any plans to do anything regarding this item.  If you are thinking of the deprecation of IAssertionRule, I did complete that a while back did not send a PR because I was not sure you would want another deprecation in v3.3, and with Issue 22 behind me now so is my motivation for deprecating the interface.

Skimming back through this, I really do not see this replacing ShouldBeEquivalentTo's behavior and I am not perceiving the value in spending time writing.  If someone valued this enough to submit a thought out PR providing an option for something like this which justifies what the functionality is needed for, there may be value in integrating it but I do not see the case for either of us to write it.

I do find my idea here about moving the Cyclic Reference detection into an IEquivalencyStep intriguing.  A quick look at the code makes me *think* it should be possible. That would add some more uniformity to the implementation of Equivalence and allow users greater extensibility, including removing the issue the OP faced with needing to modify Equivalency Validator.  However, if that is worth pursing it probably would make sense to create a new issue to track that.
